src/intelligence_layer/evaluation/benchmark/get_code.py

```diff
@@ -0,0 +1,277 @@
+"""
+The following code is licensed under the following license and taken from https://github.com/wandb/weave.
+The following modifications have been made:
+ - Adapt the type hints to be more specific to this project
+ - Change the control flow for the get_source_notebook_safe to try to only get the source from notebooks once it fails.
+
+                              Apache License
+                        Version 2.0, January 2004
+                     http://www.apache.org/licenses/
+
+TERMS AND CONDITIONS FOR USE, REPRODUCTION, AND DISTRIBUTION
```

Comment:
Could keep just the header of the license? We do not need the whole TERMS AND CONDITIONS. Also could also add the date you get this part?

Reply:
I would argue that we need the full thing (and a comments what we changed):
From https://fossa.com/blog/open-source-licenses-101-apache-license-2-0/
"
Anyone who uses open source software licensed under Apache 2.0 must include the following in their copy of the code, whether they have modified it or not:

- The original copyright notice
- A copy of the license itself
- If applicable, a statement of any significant changes made to the original code
- A copy of the NOTICE file with attribution notes (if the original library has one)
"
Point 4 should not be needed. Point 3 is up to discussion (what is "significant"?)

Reply:
Another clean suggestion from one of our PaNs was to include a NOTICE file. This also scales to future packages or source code, since we can simply mention the package and put the lincence in there

Reply:
Agree that the license needs to be there verbatim, but I would not yet do the notice file, because the format for this is quite specific. We did point 3 at the top of the license.